### ISSUE TITLE
UnicodeDecodeError or running python tests

### ISSUE THREAD

**Issue by alexander-mead** [LSSTDESC/CCL#332]
I am getting tests failing when I run the python tests using ```python setup.py test```. Strangely the number of tests that fail changes run-by-run, but has always been between 3 and 6 fails. Most of the error messages have 'UnicodeDecodeError' in them and come from 'Test mass function and supporting functions'. There is also one error which is the same as that reported in #326 . All the tests pass in C or C++, so I think this is a python problem.

OSx: 10.13.3
gsl: 2.4
Python: 3.6.4

The output of the python tests is copied and pasted below:

................................EEEEE.........E......
======================================================================
ERROR: Test mass function and supporting functions.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/Cellar/numpy/1.14.0/libexec/nose/lib/python3.6/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/Mead/Physics/CCL/tests/ccl_test_pyccl_interface.py", line 181, in check_massfunc
    assert_raises(RuntimeError, ccl.massfunc, cosmo, mhalo_scl, a, 199.)
  File "/usr/local/lib/python3.6/site-packages/numpy/testing/nose_tools/utils.py", line 1191, in assert_raises
    return nose.tools.assert_raises(*args,**kwargs)
  File "/usr/local/Cellar/python3/3.6.4_2/Frameworks/Python.framework/Versions/3.6/lib/python3.6/unittest/case.py", line 733, in assertRaises
    return context.handle('assertRaises', args, kwargs)
  File "/usr/local/Cellar/python3/3.6.4_2/Frameworks/Python.framework/Versions/3.6/lib/python3.6/unittest/case.py", line 178, in handle
    callable_obj(*args, **kwargs)
  File "/Users/Mead/Library/Python/3.6/lib/python/site-packages/pyccl/massfunction.py", line 23, in massfunc
    lib.massfunc_vec, cosmo, halo_mass, a, odelta)
  File "/Users/Mead/Library/Python/3.6/lib/python/site-packages/pyccl/pyutils.py", line 249, in _vectorize_fn4
    check(status, cosmo_in)
  File "/Users/Mead/Library/Python/3.6/lib/python/site-packages/pyccl/pyutils.py", line 18, in check
    msg = _cosmology_obj(cosmo).status_message
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xb8 in position 100: invalid start byte

======================================================================
ERROR: Test mass function and supporting functions.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/Cellar/numpy/1.14.0/libexec/nose/lib/python3.6/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/Mead/Physics/CCL/tests/ccl_test_pyccl_interface.py", line 181, in check_massfunc
    assert_raises(RuntimeError, ccl.massfunc, cosmo, mhalo_scl, a, 199.)
  File "/usr/local/lib/python3.6/site-packages/numpy/testing/nose_tools/utils.py", line 1191, in assert_raises
    return nose.tools.assert_raises(*args,**kwargs)
  File "/usr/local/Cellar/python3/3.6.4_2/Frameworks/Python.framework/Versions/3.6/lib/python3.6/unittest/case.py", line 733, in assertRaises
    return context.handle('assertRaises', args, kwargs)
  File "/usr/local/Cellar/python3/3.6.4_2/Frameworks/Python.framework/Versions/3.6/lib/python3.6/unittest/case.py", line 178, in handle
    callable_obj(*args, **kwargs)
  File "/Users/Mead/Library/Python/3.6/lib/python/site-packages/pyccl/massfunction.py", line 23, in massfunc
    lib.massfunc_vec, cosmo, halo_mass, a, odelta)
  File "/Users/Mead/Library/Python/3.6/lib/python/site-packages/pyccl/pyutils.py", line 249, in _vectorize_fn4
    check(status, cosmo_in)
  File "/Users/Mead/Library/Python/3.6/lib/python/site-packages/pyccl/pyutils.py", line 18, in check
    msg = _cosmology_obj(cosmo).status_message
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x9a in position 126: invalid start byte

======================================================================
ERROR: Test mass function and supporting functions.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/Cellar/numpy/1.14.0/libexec/nose/lib/python3.6/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/Mead/Physics/CCL/tests/ccl_test_pyccl_interface.py", line 181, in check_massfunc
    assert_raises(RuntimeError, ccl.massfunc, cosmo, mhalo_scl, a, 199.)
  File "/usr/local/lib/python3.6/site-packages/numpy/testing/nose_tools/utils.py", line 1191, in assert_raises
    return nose.tools.assert_raises(*args,**kwargs)
  File "/usr/local/Cellar/python3/3.6.4_2/Frameworks/Python.framework/Versions/3.6/lib/python3.6/unittest/case.py", line 733, in assertRaises
    return context.handle('assertRaises', args, kwargs)
  File "/usr/local/Cellar/python3/3.6.4_2/Frameworks/Python.framework/Versions/3.6/lib/python3.6/unittest/case.py", line 178, in handle
    callable_obj(*args, **kwargs)
  File "/Users/Mead/Library/Python/3.6/lib/python/site-packages/pyccl/massfunction.py", line 23, in massfunc
    lib.massfunc_vec, cosmo, halo_mass, a, odelta)
  File "/Users/Mead/Library/Python/3.6/lib/python/site-packages/pyccl/pyutils.py", line 249, in _vectorize_fn4
    check(status, cosmo_in)
  File "/Users/Mead/Library/Python/3.6/lib/python/site-packages/pyccl/pyutils.py", line 18, in check
    msg = _cosmology_obj(cosmo).status_message
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xd2 in position 106: invalid continuation byte

======================================================================
ERROR: Test mass function and supporting functions.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/Cellar/numpy/1.14.0/libexec/nose/lib/python3.6/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/Mead/Physics/CCL/tests/ccl_test_pyccl_interface.py", line 181, in check_massfunc
    assert_raises(RuntimeError, ccl.massfunc, cosmo, mhalo_scl, a, 199.)
  File "/usr/local/lib/python3.6/site-packages/numpy/testing/nose_tools/utils.py", line 1191, in assert_raises
    return nose.tools.assert_raises(*args,**kwargs)
  File "/usr/local/Cellar/python3/3.6.4_2/Frameworks/Python.framework/Versions/3.6/lib/python3.6/unittest/case.py", line 733, in assertRaises
    return context.handle('assertRaises', args, kwargs)
  File "/usr/local/Cellar/python3/3.6.4_2/Frameworks/Python.framework/Versions/3.6/lib/python3.6/unittest/case.py", line 178, in handle
    callable_obj(*args, **kwargs)
  File "/Users/Mead/Library/Python/3.6/lib/python/site-packages/pyccl/massfunction.py", line 23, in massfunc
    lib.massfunc_vec, cosmo, halo_mass, a, odelta)
  File "/Users/Mead/Library/Python/3.6/lib/python/site-packages/pyccl/pyutils.py", line 249, in _vectorize_fn4
    check(status, cosmo_in)
  File "/Users/Mead/Library/Python/3.6/lib/python/site-packages/pyccl/pyutils.py", line 18, in check
    msg = _cosmology_obj(cosmo).status_message
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x99 in position 142: invalid start byte

======================================================================
ERROR: Test mass function and supporting functions.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/Cellar/numpy/1.14.0/libexec/nose/lib/python3.6/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/Mead/Physics/CCL/tests/ccl_test_pyccl_interface.py", line 181, in check_massfunc
    assert_raises(RuntimeError, ccl.massfunc, cosmo, mhalo_scl, a, 199.)
  File "/usr/local/lib/python3.6/site-packages/numpy/testing/nose_tools/utils.py", line 1191, in assert_raises
    return nose.tools.assert_raises(*args,**kwargs)
  File "/usr/local/Cellar/python3/3.6.4_2/Frameworks/Python.framework/Versions/3.6/lib/python3.6/unittest/case.py", line 733, in assertRaises
    return context.handle('assertRaises', args, kwargs)
  File "/usr/local/Cellar/python3/3.6.4_2/Frameworks/Python.framework/Versions/3.6/lib/python3.6/unittest/case.py", line 178, in handle
    callable_obj(*args, **kwargs)
  File "/Users/Mead/Library/Python/3.6/lib/python/site-packages/pyccl/massfunction.py", line 23, in massfunc
    lib.massfunc_vec, cosmo, halo_mass, a, odelta)
  File "/Users/Mead/Library/Python/3.6/lib/python/site-packages/pyccl/pyutils.py", line 249, in _vectorize_fn4
    check(status, cosmo_in)
  File "/Users/Mead/Library/Python/3.6/lib/python/site-packages/pyccl/pyutils.py", line 18, in check
    msg = _cosmology_obj(cosmo).status_message
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xaa in position 100: invalid start byte

======================================================================
ERROR: Test top-level functions in pyccl.cls module.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/Cellar/numpy/1.14.0/libexec/nose/lib/python3.6/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/Mead/Physics/CCL/tests/ccl_test_pyccl_interface.py", line 349, in check_cls
    assert_( all_finite(ccl.angular_cl(cosmo, nc2, nc3, ell_arr)) )
  File "/Users/Mead/Library/Python/3.6/lib/python/site-packages/pyccl/cls.py", line 311, in angular_cl
    check(status, cosmo_in)
  File "/Users/Mead/Library/Python/3.6/lib/python/site-packages/pyccl/pyutils.py", line 24, in check
    raise RuntimeError("Error %s: %s" % (pyccl.core.error_types[status], msg))
RuntimeError: Error CCL_ERROR_INTEG: ccl_cls.c: ccl_angular_cl(): error integrating over k

----------------------------------------------------------------------
Ran 53 tests in 163.858s

FAILED (errors=6)

**Comment by villarrealas** [comment]
I've noted these errors show up in the Travis integrations as well, while not actually showing up as a Travis-CI failure. It seems to be primarily with Python 3.

**Comment by philbull** [comment]
Can you provide a minimal script that (sometimes) causes the same error, please?

**Comment by fjaviersanchez** [comment]
This is a known issue with python3 (see: http://lucumr.pocoo.org/2014/1/5/unicode-in-2-and-3/ and https://docs.python.org/2/howto/pyporting.html)

**Comment by alexander-mead** [comment]
The minimal script is: ```python setup.py test``` where my 'python' points to a homebrew install of python3.

**Comment by philbull** [comment]
OK, please send your Mac in the post and I'll take a look.

**Comment by philbull** [comment]
Right, thanks @fjaviersanchez. So since I'm a dinosaur who doesn't have a working Python 3.6 interpreter to hand, I can't check to see how we should be dealing with these strings. Perhaps the swig wrapper is returning them as `bytes` objects, rather than `string`? Can anyone check this?

**Comment by fjaviersanchez** [comment]
I think that if we force SWIG to return the strings as unicode should be enough

**Comment by fjaviersanchez** [comment]
https://stackoverflow.com/questions/42148068/how-can-i-make-swig-return-unicode-objects-for-strings-in-python-2

**Comment by philbull** [comment]
OK, I think I've tracked down the bug. It's pretty bizarre.

The solution was to regenerate the SWIG wrapper. It was returning a garbled string that was a mix of the proper error message, some junk characters (that were causing the unicode issue), and a docstring from a different function! Very weird. This could have been caused by a merge gone wrong or something I suppose. (The autogenerated swig files should always be regenerated from scratch before a commit, never merged.) The tests now seem to be passing.

The problem remains that unhelpful unicode errors will be raised if a garbled string gets passed from the CCL C code. This is only a problem for Python 3 because of the issues that @fjaviersanchez posted about -- Python 3 decodes strings to unicode by default, and has a default behaviour of raising errors if it finds bad characters. Python 2 just displays the ascii string, including the bad characters. (It seems that there is no easy way to sanitise strings containing bad characters in Python 3, as they are automatically decoded to unicode.) I don't think we need to worry about this for now though, as the CCL C code should always return valid strings -- it's a bug if not.
